src/stream/src/executor/top_n_new.rs

```diff
@@ -209,40 +237,167 @@ impl<S: StateStore> TopNExecutorBase for InnerTopNExecutorNew<S> {
+                    let elem_to_compare_with_middle = if self.offset > 0
+                        && ordered_pk_row <= *self.cache.low.last_key_value().unwrap().0
+                    {
+                        // If the new row is in the range of [0, offset), the largest row in
+                        // `cache.low` needs be moved to `cache.middle`
+                        // which covers the range of [offset, offset+limit)
+                        let res = self.cache.low.pop_last().unwrap();
+                        self.cache.low.insert(ordered_pk_row, row);
+                        res
+                    } else {
+                        (ordered_pk_row, row)
+                    };
+
+                    if self.cache.middle.len() < num_limit {
+                        self.cache.middle.insert(
+                            elem_to_compare_with_middle.0,
+                            elem_to_compare_with_middle.1.clone(),
+                        );
+                        res_ops.push(Op::Insert);
+                        res_rows.push(elem_to_compare_with_middle.1);
+                        continue;
+                    }
+
+                    let elem_to_compare_with_high = if elem_to_compare_with_middle.0
+                        <= *self.cache.middle.last_key_value().unwrap().0
+                    {
+                        // If the row in the range of [offset, offset+limit), the largest row in
+                        // `cache.middle` needs to be moved to `cache.high`
+                        let res = self.cache.middle.pop_last().unwrap();
+                        res_ops.push(Op::Delete);
+                        res_rows.push(res.1.clone());
+                        res_ops.push(Op::Insert);
+                        res_rows.push(elem_to_compare_with_middle.1.clone());
+                        self.cache
+                            .middle
+                            .insert(elem_to_compare_with_middle.0, elem_to_compare_with_middle.1);
+                        res
+                    } else {
+                        elem_to_compare_with_middle
+                    };
+
+                    if !self.cache.is_high_cache_full() {
+                        self.cache
+                            .high
+                            .insert(elem_to_compare_with_high.0, elem_to_compare_with_high.1);
+                    } else if elem_to_compare_with_high.0
+                        <= *self.cache.high.last_key_value().unwrap().0
+                    {
+                        self.cache.high.pop_last();
+                        self.cache
+                            .high
+                            .insert(elem_to_compare_with_high.0, elem_to_compare_with_high.1);
+                    }
                 }
 
                 Op::Delete | Op::UpdateDelete => {
-                    self.managed_state.delete(&ordered_pk_row, row, epoch)?;
+                    // First remove the row from state store
```

Comment:
This part  looks very complicated as there's a lot of `if` `else if` judgment, can we simplify it?

Reply:
Actually the theory is simple. For a delete input row, if it is larger than the largest row of [offset, offset+limit) then just remove from `high`. If the row in [offset, offset+limit), then remove it and move one row from [offset+limit, +Inf) to result set if any.